각 User가 7번의 룰렛을 돌려 획득한 점수와 상금을 기록한 데이터이다.
[데이터셋 위치(경로)](http://jhserver.dimigo.hs.kr/test16.csv)

In [1]:
import pandas as pd



In [3]:
df = pd.read_csv('http://jhserver.dimigo.hs.kr/test16.csv')


In [5]:
df.head()

User  First  Second  Third  Fourth  Fifth  Sixth  Seventh  Total     Avg  \
0  User33     10      11     30      12     11      6       29    109  15.571   
1  User38     25       1     23      35     19     17       26    146  20.857   
2  User20     35       1      1       8      5     20       11     81  11.571   
3  User12     11       5     11      35      2     10        3     77  11.000   
4  User14      4      33      2      13     35      5       26    118  16.857   

   Prize  
0   \500  
1  \1000  
2   \500  
3    NaN  
4   \500

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   User     10000 non-null  object 
 1   First    10000 non-null  int64  
 2   Second   10000 non-null  int64  
 3   Third    10000 non-null  int64  
 4   Fourth   10000 non-null  int64  
 5   Fifth    10000 non-null  int64  
 6   Sixth    10000 non-null  int64  
 7   Seventh  10000 non-null  int64  
 8   Total    10000 non-null  int64  
 9   Avg      10000 non-null  float64
 10  Prize    9677 non-null   object 
dtypes: float64(1), int64(8), object(2)
memory usage: 859.5+ KB


In [38]:
df.isnull().sum()

User         0
First        0
Second       0
Third        0
Fourth       0
Fifth        0
Sixth        0
Seventh      0
Total        0
Avg          0
Prize      323
dtype: int64

1. 주어진 데이터셋을 불러온 후 데이터셋의 자료형(dtype)의 종류와 결측값이 존재하는 컬럼과 개수를 작성하세요

In [ ]:
df = pd.read_csv('http://jhserver.dimigo.hs.kr/test16.csv')
print(df.info())
print(df.isnull().sum())

 0   User    - object
 1   First   - int64
 2   Second  - int64
 3   Third   - int64
 4   Fourth  - int64
 5   Fifth   - int64
 6   Sixth   - int64
 7   Seventh - int64
 8   Total   - int64
 9   Avg     - float64
 10  Prize   - object

 Prize 323개

In [16]:
print(df.sort_values(by='Total')
    .iloc[18])

User       User39
First           3
Second          5
Third           2
Fourth          7
Fifth          18
Sixth          17
Seventh         5
Total          57
Avg         8.143
Prize         NaN
Name: 8383, dtype: object


In [ ]:
df.sort_values(by='Total').head(19)

In [39]:
df.groupby('User').Total.sum().sort_values()[[18]]

User
User21    32002
Name: Total, dtype: int64

2. 데이터셋에서 User별 총점의 합 순으로 오름차순 정렬했을 때 본인 번호의 순위에 해당하는 User는 누구인가?(동순위가 발생한 경우 총점의 합이 높은 순으로 작성)

In [72]:
print(df.groupby('User').Total.sum().sort_values()[[17, 18, 19]])

User
User22    31975
User21    32002
User28    32303
Name: Total, dtype: int64


In [ ]:
# 24 version
print(df.groupby('User').Total.sum().sort_values()[[22, 23, 24]])

In [83]:
# 32 version
print(df.groupby('User').Total.sum().sort_values()[[30, 31, 32]])

User
User4     33812
User27    34176
User30    34274
Name: Total, dtype: int64


In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   User     10000 non-null  object 
 1   First    10000 non-null  int64  
 2   Second   10000 non-null  int64  
 3   Third    10000 non-null  int64  
 4   Fourth   10000 non-null  int64  
 5   Fifth    10000 non-null  int64  
 6   Sixth    10000 non-null  int64  
 7   Seventh  10000 non-null  int64  
 8   Total    10000 non-null  int64  
 9   Avg      10000 non-null  float64
 10  Prize    9677 non-null   object 
dtypes: float64(1), int64(8), object(2)
memory usage: 859.5+ KB


In [53]:
df.Prize = df.Prize.fillna('\\0')
df.Prize = df.Prize.apply(lambda x: int(x[1:]))

In [54]:
df.head()

User  First  Second  Third  Fourth  Fifth  Sixth  Seventh  Total     Avg  \
0  User33     10      11     30      12     11      6       29    109  15.571   
1  User38     25       1     23      35     19     17       26    146  20.857   
2  User20     35       1      1       8      5     20       11     81  11.571   
3  User12     11       5     11      35      2     10        3     77  11.000   
4  User14      4      33      2      13     35      5       26    118  16.857   

   Prize  
0    500  
1   1000  
2    500  
3      0  
4    500

In [61]:
df[df.First == 19].sort_values(by='Prize', ascending=False).head(10)

User  First  Second  Third  Fourth  Fifth  Sixth  Seventh  Total  \
6530  User14     19      35     24      22     36     34       34    204   
4294  User32     19      31     27      21     21     32       14    165   
5957   User9     19      35     33      14     27     24       13    165   
3644  User12     19      31     17      34     23     12       33    169   
1433   User1     19      35     21      25     34      5       25    164   
1552  User38     19      18     30      33     14     28       29    171   
3505  User26     19       7     29      35     25     26       24    165   
5505  User13     19      33     28      26     35     21       36    198   
1664   User3     19      33     32      30     21     33       25    193   
7788   User5     19      23     15      34     19     33       17    160   

         Avg  Prize  
6530  29.143   3000  
4294  23.571   2000  
5957  23.571   2000  
3644  24.143   2000  
1433  23.429   2000  
1552  24.429   2000  
3505  23.571   2000  
5505  28.286   2000  
1664  27.571   2000  
7788  22.857   2000

In [62]:
df[df.First == 19].sort_values(by=['Prize', 'Total'], ascending=False).iloc[5]

User       User37
First          19
Second         32
Third          31
Fourth         16
Fifth          20
Sixth          34
Seventh        30
Total         182
Avg            26
Prize        2000
Name: 8563, dtype: object

3. 데이터셋에서 첫 번째 획득 점수가 자신의 번호와 같은 게임 중에서 여섯 번째로 가장 많은 상금을 획득한 User는 누구인가?(동순위가 발생한 경우 총점의 합이 높은 경우로 작성)

In [ ]:
df.Prize = df.Prize.fillna('\\0')
df.Prize = df.Prize.apply(lambda x: int(x[1:]))
print(df[df.First == 19].sort_values(by=['Prize', 'Total'], ascending=False).iloc[5])

In [80]:
# 24 ver
print(df[df.First == 32].sort_values(by=['Prize', 'Total'], ascending=False).iloc[5])

User       User17
First          32
Second         24
Third          31
Fourth         24
Fifth          34
Sixth          25
Seventh        22
Total         192
Avg        27.429
Prize        2000
Name: 1573, dtype: object


4. Python의 Collections모듈에서 deque 컨테이너의 pop()함수와 remove() 함수의 차이저을 간략히 기술하세요

In [63]:
from collections import deque

In [64]:
a = deque([1, 2, 1, 3, 1, 4])

In [65]:
a.remove(1)

In [66]:
print(a)

deque([2, 1, 3, 1, 4])
